### PR TITLE
Implement Serpentine Helm

### PR DESCRIPTION
**Pull request by RHQ-Rusty**
This PR lays the groundwork for Barrows Gear implementation (same mechanic as for Serpentine Helm) and also implements the Serpentine Helm, with three variable modes of display: scales, percent, or both. Additionally, the list of tracked weapons was reformatted and updated in the README.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -22,7 +22,7 @@ dependencies {
 	testImplementation group: 'net.runelite', name:'client', version: runeLiteVersion
 	testImplementation group: 'net.runelite', name:'jshell', version: runeLiteVersion
 
-    testImplementation 'org.projectlombok:lombok:1.18.4'
+	testImplementation 'org.projectlombok:lombok:1.18.4'
```
Comment (RHQ-Rusty) [inline build.gradle:25]: Nothing is changed here; unsure of why it's showing like this in the diff. Shows aligned properly in-editor.

**Comment by RHQ-Rusty** [comment]
Some files also show as having every line changed; this is because I swapped them out (dcbaa015cda4204caa6026a8976f30accec4274a) and swapped them back in (b995fe54bae529ff722cb0121028e6ac4b127d5a). Nothing was changed in them.

File: src/main/java/com/weaponcharges/WeaponChargesPlugin.java [hunk 2]
```diff
@@ -214,8 +214,24 @@ private void enableDevMode()
...
 
+	private int lastDegradedHitsplatTick = -1000;
+
 	@Subscribe
 	public void onHitsplatApplied(HitsplatApplied e) {
+		// Easily Identify the Hit Type
+		Hitsplat.HitsplatType hitType = e.getHitsplat().getHitsplatType();
+		// Serpentine Helmet Logic
+		ChargedWeapon serpHelm = getEquippedChargedWeapon(EquipmentInventorySlot.HEAD);
+		if (hitType != Hitsplat.HitsplatType.VENOM || hitType != Hitsplat.HitsplatType.POISON || hitType != Hitsplat.HitsplatType.HEAL || hitType != Hitsplat.HitsplatType.DISEASE)  {
```
[the remark this answers is not in the document]
Comment (geheur) [reply src/main/java/com/weaponcharges/WeaponChargesPlugin.java:225]: also a damage_other hitsplat meets the criteria even though it shouldn't. This should be a whitelist not a blacklist.

**Review by geheur — CHANGES_REQUESTED** [review]
Some things that didn't fit in any specific place:

I don't like that the percentage thing only applies to the serp helm, it's kind of weird. but I guess it's ok for now.

You changed line endings (I assume based on the giant diffs) on a lot of files, did you really need to do that to get it to run?

The code which tracks charges is ok I guess once it's fixed. There are some weird timing things, at least with ranged attacks, but I cba doing it and I imagine you cba as well.

File: build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (RHQ-Rusty) [reply build.gradle:25]: I'm unsure of the white space change; opening both files in an editor yields the same exact display. Perhaps an invisible character?

File: src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java [hunk 3]
```diff
@@ -119,6 +122,44 @@ public void renderItemOverlay(Graphics2D graphics, int itemId, WidgetItem itemWi
...
+			charges = plugin.getCharges(chargedWeapon);
+			if (charges == null)
+			{
+				bottomText.setText("??.?%");
+			} else {
+				if (displayWhen == DisplayWhen.NEVER && !plugin.isShowChargesKeyIsDown()) return;
+				NumberFormat df = new DecimalFormat("##0.0");
+				if (config.serpentine_helm_DisplayMode() == SerpModes.SCALES) {
+					topText.setText(charges.toString());
+				} else if (config.serpentine_helm_DisplayMode() == SerpModes.PERCENT) {
```
[the remark this answers is not in the document]
Comment (RHQ-Rusty) [reply src/main/java/com/weaponcharges/WeaponChargesItemOverlay.java:138]: Where is the duplication? This is checking what display mode is set, it's a bundle of conditional logic.

File: src/test/java/com/weaponcharges/ConfigGenerator.java [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 public class ConfigGenerator
 {
-	static int i = 0;
+	static int i = -1;
```
[the remark this answers is not in the document]
Comment (RHQ-Rusty) [reply src/test/java/com/weaponcharges/ConfigGenerator.java:7]: A necessary change to get the updated ConfigGenerator to work. This is tested and causes no issues.

**Comment by RHQ-Rusty** [comment]
Most of the white-space, line-ending, etc. errors can be attributed to the removal and re-upload of the files; if this is that big of an issue I can recreate the fork and leave the source files as they were. This would only show my changes. Let me know what you would like me to do before I continue with the PR.

**Pull request closed without merging**